File: src/main/java/io/curiousoft/ijudi/ordermanagement/model/Order.java
package io.curiousoft.ijudi.ordermanagement.model;

import io.curiousoft.ijudi.ordermanagement.validator.ValidDeliveryInfo;
import io.curiousoft.ijudi.ordermanagement.validator.ValidOrderType;
import org.springframework.data.mongodb.core.mapping.Document;

import javax.validation.Valid;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.PositiveOrZero;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.time.ZoneOffset;

@ValidOrderType
@ValidDeliveryInfo
@Document
public class Order extends BaseModel {

    @NotNull(message = "order stage is not valid")
    private OrderStage stage;

    private ShippingData shippingData;
    @Valid
    @NotNull(message = "order basket is not valid")
    private Basket basket;
    @NotBlank(message = "order customer is not valid")
    private String customerId;
    @NotBlank(message = "order shop is not valid")
    private String shopId;
    @NotBlank(message = "order description is not valid")
    private String description;
    private PaymentType paymentType;
    @NotNull(message = "order type is not valid")
    private OrderType orderType;
    private boolean hasVat;
    private boolean shopPaid;
    @PositiveOrZero(message = "service fee was modified")
    private double serviceFee;
    private boolean messengerPaid;
    private boolean smsSentToShop;
    private boolean smsSentToAdmin;
    private boolean freeDelivery;
    private double minimumDepositAllowedPerc;

    public void setStage(OrderStage stage) {
        this.stage = stage;
    }

    public OrderStage getStage() {
        return stage;
    }

    public ShippingData getShippingData() {
        return shippingData;
    }

    public void setShippingData(ShippingData shippingData) {
        this.shippingData = shippingData;
    }

    public Basket getBasket() {
        return basket;
    }

    public void setBasket(Basket basket) {
        this.basket = basket;
    }

    public String getCustomerId() {
        return customerId;
    }

    public void setCustomerId(String customerId) {
        this.customerId = customerId;
    }

    public String getShopId() {
        return shopId;
    }

    public void setShopId(String shopId) {
        this.shopId = shopId;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public PaymentType getPaymentType() {
        return paymentType;
    }

    public void setPaymentType(PaymentType paymentType) {
        this.paymentType = paymentType;
    }

    public double getTotalAmount() {
        return BigDecimal.valueOf(serviceFee + basket.getItems().stream()
                .mapToDouble(BasketItem::getTotalPrice).sum() + (!freeDelivery && shippingData != null ? shippingData.getFee() : 0))
                .setScale(2, RoundingMode.HALF_EVEN)
                .doubleValue();
    }

    public OrderType getOrderType() {
        return orderType;
    }

    public void setOrderType(OrderType orderType) {
        this.orderType = orderType;
    }

    public boolean getHasVat() {
        return hasVat;
    }

    public void setHasVat(boolean hasVat) {
        this.hasVat = hasVat;
    }

    public boolean getShopPaid() {
        return shopPaid;
    }

    public void setShopPaid(boolean shopPaid) {
        this.shopPaid = shopPaid;
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof Order && getId().equals(((Order) obj).getId());
    }

    public double getBasketAmount() {
        return BigDecimal.valueOf(basket.getItems().stream()
                .mapToDouble(BasketItem::getTotalPrice).sum())
                .setScale(2, RoundingMode.HALF_EVEN)
                .doubleValue();
    }

    public double getServiceFee() {
        return serviceFee;
    }

    public void setServiceFee(double serviceFee) {
        this.serviceFee = serviceFee;
    }

    public boolean getMessengerPaid() {
        return messengerPaid;
    }

    public void setMessengerPaid(boolean messengerPaid) {
        this.messengerPaid = messengerPaid;
    }

    public static String generateId() {
        return (LocalDateTime.now().toEpochSecond(ZoneOffset.UTC)) + "";
    }

    public void setSmsSentToShop(boolean smsSentToShop) {
        this.smsSentToShop = smsSentToShop;
    }

    public boolean getSmsSentToShop() {
        return smsSentToShop;
    }

    public void setSmsSentToAdmin(boolean smsSentToAdmin) {
        this.smsSentToAdmin = smsSentToAdmin;
    }

    public boolean getSmsSentToAdmin() {
        return smsSentToAdmin;
    }

    public boolean getFreeDelivery() {
        return freeDelivery;
    }

    public void setFreeDelivery(boolean freeDelivery) {
        this.freeDelivery = freeDelivery;
    }

    public double getDepositAmount() {
        return BigDecimal.valueOf(serviceFee
                + basket.getItems().stream().mapToDouble(BasketItem::getTotalPrice).sum() * minimumDepositAllowedPerc
                + (!freeDelivery && shippingData != null ? shippingData.getFee() : 0))
                .setScale(2, RoundingMode.HALF_EVEN)
                .doubleValue();
    }

    public void setMinimumDepositAllowedPerc(double minimumDepositAllowedPerc) {
        this.minimumDepositAllowedPerc = minimumDepositAllowedPerc;
    }

    public double getMinimumDepositAllowedPerc() {
        return minimumDepositAllowedPerc;
    }
}

File: izinga-commons/src/main/kotlin/io/curiousoft/izinga/commons/model/Order.kt
package io.curiousoft.izinga.commons.model

import io.curiousoft.izinga.commons.validator.ValidDeliveryInfo
import io.curiousoft.izinga.commons.validator.ValidOrderType
import org.springframework.data.mongodb.core.mapping.Document
import java.lang.StringBuilder
import java.math.BigDecimal
import java.math.RoundingMode
import java.security.SecureRandom
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.*
import javax.validation.Valid
import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull

@ValidOrderType
@ValidDeliveryInfo
@Document
class Order : BaseModel() {
    var paymentTypesAllowed: MutableList<PaymentType>? = null
    var stage: @NotNull(message = "order stage is not valid") OrderStage? = null
    var shippingData: ShippingData? = null
    lateinit var basket: @Valid @NotNull(message = "order basket is not valid") Basket
    var customerId: @NotBlank(message = "order customer is not valid") String? = null
    lateinit var shopId: @NotBlank(message = "order shop is not valid") String
    var description: @NotBlank(message = "order description is not valid") String? = null
    var paymentType: PaymentType? = null
    var orderType: @NotNull(message = "order type is not valid") OrderType? = null
    var hasVat = false
    var shopPaid = false
    var shopPaidDate: Date? = null
    var serviceFee = 0.0
    var tip: Double? = null
    var messengerPaid = false
    var messengerPaidDate: Date? = null
    var smsSentToShop = false
    var smsSentToAdmin = false
    var freeDelivery = false
    var minimumDepositAllowedPerc = 0.0
    var payoutCreated = false
    var documents: Set<DocumentAttachment>? = null
    val totalAmount: Double
        get() = BigDecimal.valueOf(
            serviceFee + basket.totalPrice + (!freeDelivery && shippingData != null).isTrue({shippingData?.fee!!}) { 0.00 }
        )
        .setScale(2, RoundingMode.HALF_EVEN)
        .toDouble()

    override fun equals(obj: Any?): Boolean {
        return obj is Order && id == obj.id
    }

    val basketAmount: Double
        get() = BigDecimal.valueOf(basket.items
            .sumOf { obj: BasketItem -> obj.totalPrice }
        )
            .setScale(2, RoundingMode.HALF_EVEN)
            .toDouble()
    val depositAmount: Double
        get() = BigDecimal.valueOf(serviceFee
                + basket.items.sumOf { obj: BasketItem -> obj.totalPrice } * minimumDepositAllowedPerc + (!freeDelivery && shippingData != null).isTrue({ shippingData?.fee!! }) { 0.00 }
        )
            .setScale(2, RoundingMode.HALF_EVEN)
            .toDouble()
}

fun <R> Boolean.isTrue(isTrue: () -> R, isFalse : () -> R): R = if (this) isTrue.invoke() else isFalse.invoke()

fun generateId(): String {
    val random = SecureRandom()
    val characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    val sb = StringBuilder()
    while (sb.length < 5) {
        sb.append(characters[random.nextInt(characters.length)])
    }
    return sb.substring(0, 5).uppercase(Locale.getDefault())
}